ext/mysqli/mysqli_api.c

```diff
@@ -1698,31 +1708,20 @@ PHP_FUNCTION(mysqli_options)
 	}
 #endif
 	expected_type = mysqli_options_get_option_zval_type(mysql_option);
-	if (expected_type != Z_TYPE_P(mysql_value)) {
-		switch (expected_type) {
-			case IS_STRING:
-				if (!try_convert_to_string(mysql_value)) {
-					RETURN_THROWS();
-				}
-				break;
-			case IS_LONG:
-				convert_to_long_ex(mysql_value);
-				break;
-			default:
-				break;
+	if (expected_type == IS_STRING) {
+		if (!mysql_value_str) {
+			zend_argument_type_error(3, "must be of type string for the chosen option");
+			RETURN_THROWS();
 		}
-	}
-	switch (expected_type) {
-		case IS_STRING:
-			ret = mysql_options(mysql->mysql, mysql_option, Z_STRVAL_P(mysql_value));
-			break;
-		case IS_LONG:
-			l_value = Z_LVAL_P(mysql_value);
-			ret = mysql_options(mysql->mysql, mysql_option, (char *)&l_value);
-			break;
-		default:
-			ret = 1;
-			break;
+		ret = mysql_options(mysql->mysql, mysql_option, ZSTR_VAL(mysql_value_str));
+	} else if (expected_type == IS_LONG) {
+		if (mysql_value_str) {
+			zend_argument_type_error(3, "must be of type int for the chosen option");
+			RETURN_THROWS();
+		}
+		ret = mysql_options(mysql->mysql, mysql_option, (char *) &mysql_value_long);
+	} else {
+		ret = 1;
```

Comment:
I'm a bit unsure about this one. If we consider a call like `mysqli_options($link, MYSQLI_OPT_CONNECT_TIMEOUT, '100')`, this was previously valid and the `'100'` was interpreted as `100`. Now it will throw instead. That is, it's now necessary to match the type of the option, rather than the option forcing the type of the value.

Reply:
Yeah, it's a valid critique against the (former) implementation. Now, I've just pushed a change which tries to address this by trying to convert the type to the expected one if necessary. I like this solution, but I'm curious whether you see any problem with this approach.

Reply:
Not a fan of this. This has made the implementation more complicated than it was before.

I think it is better to keep this as a mixed argument. The fact that only int|string is accepted is rather incidential imho (e.g. there's no reason why in the future an option using float or bool value couldn't be added).

Reply:
I reverted all these changes (from mysqli and xml both), but stayed with the `string|int` PHPDoc type hints.